File: Upnp/Device/Services/Cpp/Core/DvLinnCoUkConfiguration1.cpp
#include "DvLinnCoUkConfiguration1.h"
#include <ZappTypes.h>
#include <Core/DvInvocationResponse.h>
#include <Service.h>
#include <FunctorDviInvocation.h>

using namespace Zapp;

TBool DvProviderLinnCoUkConfiguration1::SetPropertyConfigurationXml(const Brx& aValue)
{
    return SetPropertyString(*iPropertyConfigurationXml, aValue);
}

void DvProviderLinnCoUkConfiguration1::GetPropertyConfigurationXml(Brhz& aValue)
{
    aValue.Set(iPropertyConfigurationXml->Value());
}

TBool DvProviderLinnCoUkConfiguration1::SetPropertyParameterXml(const Brx& aValue)
{
    return SetPropertyString(*iPropertyParameterXml, aValue);
}

void DvProviderLinnCoUkConfiguration1::GetPropertyParameterXml(Brhz& aValue)
{
    aValue.Set(iPropertyParameterXml->Value());
}

DvProviderLinnCoUkConfiguration1::DvProviderLinnCoUkConfiguration1(DvDevice& aDevice)
    : DvProvider(aDevice.Device(), "linn.co.uk", "Configuration", 1)
{
    
    iPropertyConfigurationXml = new PropertyString(new ParameterString("ConfigurationXml"));
    iService->AddProperty(iPropertyConfigurationXml); // passes ownership
    iPropertyParameterXml = new PropertyString(new ParameterString("ParameterXml"));
    iService->AddProperty(iPropertyParameterXml); // passes ownership
}

void DvProviderLinnCoUkConfiguration1::EnableActionConfigurationXml()
{
    Zapp::Action* action = new Zapp::Action("ConfigurationXml");
    action->AddOutputParameter(new ParameterRelated("aConfigurationXml", *iPropertyConfigurationXml));
    FunctorDviInvocation functor = MakeFunctorDviInvocation(*this, &DvProviderLinnCoUkConfiguration1::DoConfigurationXml);
    iService->AddAction(action, functor);
}

void DvProviderLinnCoUkConfiguration1::EnableActionParameterXml()
{
    Zapp::Action* action = new Zapp::Action("ParameterXml");
    action->AddOutputParameter(new ParameterRelated("aParameterXml", *iPropertyParameterXml));
    FunctorDviInvocation functor = MakeFunctorDviInvocation(*this, &DvProviderLinnCoUkConfiguration1::DoParameterXml);
    iService->AddAction(action, functor);
}

void DvProviderLinnCoUkConfiguration1::EnableActionSetParameter()
{
    Zapp::Action* action = new Zapp::Action("SetParameter");
    action->AddInputParameter(new ParameterString("aTarget"));
    action->AddInputParameter(new ParameterString("aName"));
    action->AddInputParameter(new ParameterString("aValue"));
    FunctorDviInvocation functor = MakeFunctorDviInvocation(*this, &DvProviderLinnCoUkConfiguration1::DoSetParameter);
    iService->AddAction(action, functor);
}

void DvProviderLinnCoUkConfiguration1::DoConfigurationXml(IDviInvocation& aInvocation, TUint aVersion)
{
    aInvocation.InvocationReadStart();
    aInvocation.InvocationReadEnd();
    InvocationResponse resp(aInvocation);
    InvocationResponseString respaConfigurationXml(aInvocation, "aConfigurationXml");
    ConfigurationXml(resp, aVersion, respaConfigurationXml);
}

void DvProviderLinnCoUkConfiguration1::DoParameterXml(IDviInvocation& aInvocation, TUint aVersion)
{
    aInvocation.InvocationReadStart();
    aInvocation.InvocationReadEnd();
    InvocationResponse resp(aInvocation);
    InvocationResponseString respaParameterXml(aInvocation, "aParameterXml");
    ParameterXml(resp, aVersion, respaParameterXml);
}

void DvProviderLinnCoUkConfiguration1::DoSetParameter(IDviInvocation& aInvocation, TUint aVersion)
{
    aInvocation.InvocationReadStart();
    Brhz aTarget;
    aInvocation.InvocationReadString("aTarget", aTarget);
    Brhz aName;
    aInvocation.InvocationReadString("aName", aName);
    Brhz aValue;
    aInvocation.InvocationReadString("aValue", aValue);
    aInvocation.InvocationReadEnd();
    InvocationResponse resp(aInvocation);
    SetParameter(resp, aVersion, aTarget, aName, aValue);
}

void DvProviderLinnCoUkConfiguration1::ConfigurationXml(IInvocationResponse& /*aResponse*/, TUint /*aVersion*/, IInvocationResponseString& /*aaConfigurationXml*/)
{
    ASSERTS();
}

void DvProviderLinnCoUkConfiguration1::ParameterXml(IInvocationResponse& /*aResponse*/, TUint /*aVersion*/, IInvocationResponseString& /*aaParameterXml*/)
{
    ASSERTS();
}

void DvProviderLinnCoUkConfiguration1::SetParameter(IInvocationResponse& /*aResponse*/, TUint /*aVersion*/, const Brx& /*aaTarget*/, const Brx& /*aaName*/, const Brx& /*aaValue*/)
{
    ASSERTS();
}

File: Upnp/Device/Services/Cs/DvLinnCoUkConfiguration1.cs
using System;
using System.Runtime.InteropServices;
using System.Text;
using System.Collections.Generic;
using Zapp.Core;

namespace Zapp.Device.Providers
{
    public interface IDvProviderLinnCoUkConfiguration1 : IDisposable
    {

        /// <summary>
        /// Set the value of the ConfigurationXml property
        /// </summary>
        /// <param name="aValue">New value for the property</param>
        /// <returns>true if the value has been updated; false if aValue was the same as the previous value</returns>
        bool SetPropertyConfigurationXml(string aValue);

        /// <summary>
        /// Get a copy of the value of the ConfigurationXml property
        /// </summary>
        /// <param name="aValue">Property's value will be copied here</param>
        string PropertyConfigurationXml();

        /// <summary>
        /// Set the value of the ParameterXml property
        /// </summary>
        /// <param name="aValue">New value for the property</param>
        /// <returns>true if the value has been updated; false if aValue was the same as the previous value</returns>
        bool SetPropertyParameterXml(string aValue);

        /// <summary>
        /// Get a copy of the value of the ParameterXml property
        /// </summary>
        /// <param name="aValue">Property's value will be copied here</param>
        string PropertyParameterXml();
        
    }
    /// <summary>
    /// Provider for the linn.co.uk:Configuration:1 UPnP service
    /// </summary>
    public class DvProviderLinnCoUkConfiguration1 : DvProvider, IDisposable, IDvProviderLinnCoUkConfiguration1
    {
        private GCHandle iGch;
        private ActionDelegate iDelegateConfigurationXml;
        private ActionDelegate iDelegateParameterXml;
        private ActionDelegate iDelegateSetParameter;
        private PropertyString iPropertyConfigurationXml;
        private PropertyString iPropertyParameterXml;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="aDevice">Device which owns this provider</param>
        protected DvProviderLinnCoUkConfiguration1(DvDevice aDevice)
            : base(aDevice, "linn-co-uk", "Configuration", 1)
        {
            iGch = GCHandle.Alloc(this);
            List<String> allowedValues = new List<String>();
            iPropertyConfigurationXml = new PropertyString(new ParameterString("ConfigurationXml", allowedValues));
            AddProperty(iPropertyConfigurationXml);
            iPropertyParameterXml = new PropertyString(new ParameterString("ParameterXml", allowedValues));
            AddProperty(iPropertyParameterXml);
        }

        /// <summary>
        /// Set the value of the ConfigurationXml property
        /// </summary>
        /// <param name="aValue">New value for the property</param>
        /// <returns>true if the value has been updated; false if aValue was the same as the previous value</returns>
        public bool SetPropertyConfigurationXml(string aValue)
        {
            return SetPropertyString(iPropertyConfigurationXml, aValue);
        }

        /// <summary>
        /// Get a copy of the value of the ConfigurationXml property
        /// </summary>
        /// <returns>The value of the property</returns>
        public string PropertyConfigurationXml()
        {
            return iPropertyConfigurationXml.Value();
        }

        /// <summary>
        /// Set the value of the ParameterXml property
        /// </summary>
        /// <param name="aValue">New value for the property</param>
        /// <returns>true if the value has been updated; false if aValue was the same as the previous value</returns>
        public bool SetPropertyParameterXml(string aValue)
        {
            return SetPropertyString(iPropertyParameterXml, aValue);
        }

        /// <summary>
        /// Get a copy of the value of the ParameterXml property
        /// </summary>
        /// <returns>The value of the property</returns>
        public string PropertyParameterXml()
        {
            return iPropertyParameterXml.Value();
        }

        /// <summary>
        /// Signal that the action ConfigurationXml is supported.
        /// </summary>
        /// <remarks>The action's availability will be published in the device's service.xml.
        /// DoConfigurationXml must be overridden if this is called.</remarks>
        protected unsafe void EnableActionConfigurationXml()
        {
            Zapp.Core.Action action = new Zapp.Core.Action("ConfigurationXml");
            action.AddOutputParameter(new ParameterRelated("aConfigurationXml", iPropertyConfigurationXml));
            iDelegateConfigurationXml = new ActionDelegate(DoConfigurationXml);
            EnableAction(action, iDelegateConfigurationXml, GCHandle.ToIntPtr(iGch));
        }

        /// <summary>
        /// Signal that the action ParameterXml is supported.
        /// </summary>
        /// <remarks>The action's availability will be published in the device's service.xml.
        /// DoParameterXml must be overridden if this is called.</remarks>
        protected unsafe void EnableActionParameterXml()
        {
            Zapp.Core.Action action = new Zapp.Core.Action("ParameterXml");
            action.AddOutputParameter(new ParameterRelated("aParameterXml", iPropertyParameterXml));
            iDelegateParameterXml = new ActionDelegate(DoParameterXml);
            EnableAction(action, iDelegateParameterXml, GCHandle.ToIntPtr(iGch));
        }

        /// <summary>
        /// Signal that the action SetParameter is supported.
        /// </summary>
        /// <remarks>The action's availability will be published in the device's service.xml.
        /// DoSetParameter must be overridden if this is called.</remarks>
        protected unsafe void EnableActionSetParameter()
        {
            Zapp.Core.Action action = new Zapp.Core.Action("SetParameter");
            List<String> allowedValues = new List<String>();
            action.AddInputParameter(new ParameterString("aTarget", allowedValues));
            action.AddInputParameter(new ParameterString("aName", allowedValues));
            action.AddInputParameter(new ParameterString("aValue", allowedValues));
            iDelegateSetParameter = new ActionDelegate(DoSetParameter);
            EnableAction(action, iDelegateSetParameter, GCHandle.ToIntPtr(iGch));
        }

        /// <summary>
        /// ConfigurationXml action.
        /// </summary>
        /// <remarks>Will be called when the device stack receives an invocation of the
        /// ConfigurationXml action for the owning device.
        ///
        /// Must be implemented iff EnableActionConfigurationXml was called.</remarks>
        /// <param name="aVersion">Version of the service being requested (will be <= the version advertised)</param>
        /// <param name="aaConfigurationXml"></param>
        protected virtual void ConfigurationXml(uint aVersion, out string aaConfigurationXml)
        {
            throw (new ActionDisabledError());
        }

        /// <summary>
        /// ParameterXml action.
        /// </summary>
        /// <remarks>Will be called when the device stack receives an invocation of the
        /// ParameterXml action for the owning device.
        ///
        /// Must be implemented iff EnableActionParameterXml was called.</remarks>
        /// <param name="aVersion">Version of the service being requested (will be <= the version advertised)</param>
        /// <param name="aaParameterXml"></param>
        protected virtual void ParameterXml(uint aVersion, out string aaParameterXml)
        {
            throw (new ActionDisabledError());
        }

        /// <summary>
        /// SetParameter action.
        /// </summary>
        /// <remarks>Will be called when the device stack receives an invocation of the
        /// SetParameter action for the owning device.
        ///
        /// Must be implemented iff EnableActionSetParameter was called.</remarks>
        /// <param name="aVersion">Version of the service being requested (will be <= the version advertised)</param>
        /// <param name="aaTarget"></param>
        /// <param name="aaName"></param>
        /// <param name="aaValue"></param>
        protected virtual void SetParameter(uint aVersion, string aaTarget, string aaName, string aaValue)
        {
            throw (new ActionDisabledError());
        }

        private static unsafe int DoConfigurationXml(IntPtr aPtr, IntPtr aInvocation, uint aVersion)
        {
            GCHandle gch = GCHandle.FromIntPtr(aPtr);
            DvProviderLinnCoUkConfiguration1 self = (DvProviderLinnCoUkConfiguration1)gch.Target;
            DvInvocation invocation = new DvInvocation(aInvocation);
            string aConfigurationXml;
            try
            {
                invocation.ReadStart();
                invocation.ReadEnd();
                self.ConfigurationXml(aVersion, out aConfigurationXml);
            }
            catch (ActionError)
            {
                invocation.ReportError(501, "Invalid XML"); ;
                return -1;
            }
            catch (ActionDisabledError)
            {
                invocation.ReportError(501, "Action not implemented"); ;
                return -1;
            }
            catch (PropertyUpdateError)
            {
                invocation.ReportError(501, "Invalid XML"); ;
                return -1;
            }
            try
            {
                invocation.WriteStart();
                invocation.WriteString("aConfigurationXml", aConfigurationXml);
                invocation.WriteEnd();
            }
            catch (ActionError)
            {
                return -1;
            }
            return 0;
        }

        private static unsafe int DoParameterXml(IntPtr aPtr, IntPtr aInvocation, uint aVersion)
        {
            GCHandle gch = GCHandle.FromIntPtr(aPtr);
            DvProviderLinnCoUkConfiguration1 self = (DvProviderLinnCoUkConfiguration1)gch.Target;
            DvInvocation invocation = new DvInvocation(aInvocation);
            string aParameterXml;
            try
            {
                invocation.ReadStart();
                invocation.ReadEnd();
                self.ParameterXml(aVersion, out aParameterXml);
            }
            catch (ActionError)
            {
                invocation.ReportError(501, "Invalid XML"); ;
                return -1;
            }
            catch (ActionDisabledError)
            {
                invocation.ReportError(501, "Action not implemented"); ;
                return -1;
            }
            catch (PropertyUpdateError)
            {
                invocation.ReportError(501, "Invalid XML"); ;
                return -1;
            }
            try
            {
                invocation.WriteStart();
                invocation.WriteString("aParameterXml", aParameterXml);
                invocation.WriteEnd();
            }
            catch (ActionError)
            {
                return -1;
            }
            return 0;
        }

        private static unsafe int DoSetParameter(IntPtr aPtr, IntPtr aInvocation, uint aVersion)
        {
            GCHandle gch = GCHandle.FromIntPtr(aPtr);
            DvProviderLinnCoUkConfiguration1 self = (DvProviderLinnCoUkConfiguration1)gch.Target;
            DvInvocation invocation = new DvInvocation(aInvocation);
            string aTarget;
            string aName;
            string aValue;
            try
            {
                invocation.ReadStart();
                aTarget = invocation.ReadString("aTarget");
                aName = invocation.ReadString("aName");
                aValue = invocation.ReadString("aValue");
                invocation.ReadEnd();
                self.SetParameter(aVersion, aTarget, aName, aValue);
            }
            catch (ActionError)
            {
                invocation.ReportError(501, "Invalid XML"); ;
                return -1;
            }
            catch (ActionDisabledError)
            {
                invocation.ReportError(501, "Action not implemented"); ;
                return -1;
            }
            catch (PropertyUpdateError)
            {
                invocation.ReportError(501, "Invalid XML"); ;
                return -1;
            }
            try
            {
                invocation.WriteStart();
                invocation.WriteEnd();
            }
            catch (ActionError)
            {
                return -1;
            }
            return 0;
        }

        /// <summary>
        /// Must be called for each class instance.  Must be called before Core.Library.Close().
        /// </summary>
        public void Dispose()
        {
            DoDispose();
            GC.SuppressFinalize(this);
        }

        ~DvProviderLinnCoUkConfiguration1()
        {
            DoDispose();
        }

        private void DoDispose()
        {
            lock (this)
            {
                if (iHandle == IntPtr.Zero)
                {
                    return;
                }
                DisposeProvider();
                iHandle = IntPtr.Zero;
            }
            iGch.Free();
        }
    }
}

